Provide a Logger hierarchy, and inheritance of default fields

DavidPineauScality opened the pull request:
This PR provides a new important feature and API for our logging evolutions: A hierarchy between loggers (including request loggers).

The feature is mostly concentrated in the first commit, the DefaultFields class and its tests. It provides the old "default fields" concept, as seen previously in the RequestLogger; with the addition of a notion of hierarchy (1 parent, multiple children) that implies a default fields inheritance from the parent to the children.
This way, we can dynamically update the default fields included by a child logger with no user interaction.

The computational cost is reduced by a caching of the hierarchy's aggregated default fields.


Note: This is currently WIP, and still lacks a Readme documentation of the design.

Commit:
FT: Provide the DefaultFields class

This new utility class provides two concepts:
 - 'Default fields' that already existed within the RequestLogger
 - 'Hierarchy relationships', with override capability and automatic
    cache/update of the aggregation of the default fields provided by the whole
    hierarchy

This class is aimed at being used by other pieces of werelogs, and is provided
with its behavioral tests.

Commit:
RF: Use DefaultFields in RequestLogger.js

Replace the default fields logic from RequestLogger.js (RequestLogger and
EndLogger classes) by the use and extension of the newly written DefaultFields
class.

Commit:
FT: Use DefaultFields in Logger.js

Provide the DefaultFields class's logic to the Logger class.
This will enable new features such as linking Module loggers together to
provide a more consistent logging scheme.

Commit:
feat(hierarchy): add backwards/forward compat

Warning: An important feature update is provided, and as such, we should
clearly show this through a minor version update, but we are not
following semantic versionning, thus cannot hint at that. The current
feature is deprecating some function prototypes, as the API is bound to
change in a way that breaks backwards compatibility. We're currently
maintaining the backwards compatibility through a deprecated version of
the functions, but this should disappear from the API starting at the
next major version.

This commit also introduces the backward/forward compatibility for the
Logger's constructor, as well as the additional tests validating the
forward-compatibility of the API.

Commit:
FT: Add a parent/child relationship between Logger and RequestLogger

This feature was enabled through the use of the DefaultFields class, that
provides a relationship between its inheritors. This way, the RequestLogger now
includes the Module Logger's defaults fields in a way that is dynamic and can
be updated at runtime, instead of only at object instanciation time.

ThibaultRiviere commented on lib/DefaultFields.js:
```diff
@@ -0,0 +1,207 @@
+     *
+     * @return {undefined}
+     */
+    _dfNotifyChildren() {
```
Typo: s/dfNotifyChildren/dfNotifyChildrens/

DavidPineauScality replied:
Children is the plural of Child already :)

ThibaultRiviere replied:
My english again ^^

ThibaultRiviere commented on lib/Logger.js:
```diff
@@ -61,7 +70,7 @@ class Logger {
                                        Config.level, Config.dump, Config.end,
                                        uids);
-        rLog.addDefaultFields({ name: this.name });
+        rLog.setParent(this);
```
That will not add a lot of computation for each requests?

DavidPineauScality replied:
setting a parent is only triggerring the "transmission" of the parent's `precomputedFields` to the child, thus computing the `precomputedFields` of the child (here the `rLog`).
As such, this is not much heavier than the previous methods.

alexandre-merle replied:
do we loose the name info here ? :)

DavidPineauScality replied:
No, as it is now added to the default fields at either https://github.com/scality/werelogs/pull/86/files/3cd4272a971be31f77356bc3c0c3faa5bdfe67da#diff-87f1df4f7e51d77fc7d8943721be6631R40 or https://github.com/scality/werelogs/pull/86/files/3cd4272a971be31f77356bc3c0c3faa5bdfe67da#diff-87f1df4f7e51d77fc7d8943721be6631R44
The `name` field will thus be inherited through the DefaultFields properties.

alexandre-merle replied:
oh right, my bad :)

ThibaultRiviere commented on lib/Logger.js:
```diff
@@ -77,13 +86,15 @@ class Logger {
     _doLog(levelName, msg, data) {
         const sLogger = Config.logger;
-        const finalData = { name: this.name, time: Date.now() };
+        const finalData = Object.assign({},
```
Should we compute the finalData after checking the shouldLog avoiding lot of useless Object.assign ?
btw: why not using ```objectCopy``` from ```utils```?

DavidPineauScality replied:
Good point about using objectCopy.
As for computing the finalData, I need to compute it at the logging time, whether it's visible or not, since the fields may change in the meantime. There would be no meaning in logging fields "more recent" than the log entry itself. As there is no way to tie "default fields" and "log entry" in the timeline, this is the only option available to us.

DavidPineauScality replied:
- [ ] Use objectCopy

ThibaultRiviere reviewed (commented):
Clean work nice job .

DavidPineauScality commented on lib/DefaultFields.js:
```diff
@@ -0,0 +1,207 @@
+     * @return {undefined}
+     */
+    _dfUnregisterChild(toRemove) {
+        const filtered = this.children.filter(child => child !== toRemove);
```
I'd need to know the index of the entry before doing that. This was the shortest method I knew of, but if there are measurable negative performance impacts, I can change it.

Commit:
FT: DefaultFields: Add a memory management helper

Due to the DefaultFields's design, we setup cyclic references that a managed
language such as Node.JS has no way of solving to release the associated
memory.

As such, this commit provides an helper function that allows breaking the
references and can be called manually, in addition to being called
automatically from the end logger's logging methods.

This choice was pushed by the analysis of the possible alternatives (coming
close to naught), and the context accompanying the feature, such as the fact
the DefaultFields class is aimed at providing relationships management between
various Module loggers and the Request Loggers:
 1. The RequestLogger has a lifetime contained within that of its natural
    parent, the Module Logger
 2. The Module Logger has a lifetime matching that of the program's, since it
    is created once for all at the intialization of the associated module
As such, it was deemed acceptable to never release the Module Logger's memory,
and automate the RequestLogger's memory release through the use of the destroy
function at the end of their lifespan: The call of the endlogger's logging
method (that should be called only once, and called as the last logging entry
of the RequestLogger, thus meaning that the RequestLogger should be released
right after that, and not used anymore).

TL;DR; OOP design is shit in memory-managed languages, so here's our own shit
to work around it, and make the memory leak issue "not that bad".

alexandre-merle commented on lib/DefaultFields.js:
```diff
@@ -0,0 +1,226 @@
+        this.children.forEach(child => {
+            child.setParent(null);
+        });
+        this.setParent(null);
```
cleanup also the children array at this point can be a good idea :)

DavidPineauScality replied:
Indeed !
- [ ] Cleanup `this.children` on `Destroy()`

alexandre-merle commented on lib/DefaultFields.js:
```diff
@@ -0,0 +1,226 @@
+     * @return {DefaultFields|null} The previous parent
+     */
+    setParent(parent) {
+        const oldParent = this.parent;
```
checking if this.parent is equal to parent can improve some cases :)

edit: This also add the posibility to remove the parent of the child when ```_dfUnregisterChild``` is call from the parent :)

DavidPineauScality replied:
I understand how it would improve some cases (although admittedly minor), but I do not really see any use to unsetting the relationship from the parent ? I mean it could already be done through `child.setParent(null)`, right ?
 - [ ] Check for parent identifty to avoid doing unnecessary operations

alexandre-merle replied:
yes, but in case of a remove child directly, the parent is not reset :)

DavidPineauScality replied:
Indeed, as the `_dfRemoveChild()` function was not intended to be called by anything other than internal mechanisms. As such, what you suggests seems to be a protection of sorts against circular dependency between the two functions (each calling the other) to avoid unintended  infinite recursion.
If you think it's useful to do this, I can, although I didn't see much of a point when writing this code.

alexandre-merle replied:
Yes, true, may be not necessary, if internally we not going to call this method directly :)

DavidPineauScality replied:
I'll refrain from doing this then :)

alexandre-merle commented on lib/Logger.js:
```diff
@@ -4,15 +4,21 @@ const LogLevel = require('./LogLevel.js');
 const RequestLogger = require('./RequestLogger.js');
 const unserializeUids = require('./Utils.js').unserializeUids;
 const Config = require('./Config.js');
+const DefaultFields = require('./DefaultFields.js');
```
```./DefaultFields``` :)

DavidPineauScality replied:
Well, all the code within werelogs adds the `.js`, so I don't plan on creating inconsistency with this PR. Can be a later cleanup :)

alexandre-merle replied:
okay :)

alexandre-merle commented on tests/unit/Logger.js:
```diff
@@ -256,3 +256,81 @@ describe('Werelogs Module-level Logger can log as specified by the log level', (
+        const logger = new Logger(defaultFields, loggerConfig);
+        logger.addDefaultFields(clientInfo);
+        logger.info('test message');
+        assert.strictEqual(clientInfo.clientIP,
```
the expected key should be at the right of the call :)

DavidPineauScality replied:
Indeed.
 - [ ] Check the order of the assert parameters (expected on the right)

MathieuCassagne commented on lib/DefaultFields.js:
```diff
@@ -0,0 +1,226 @@
+     * @return {undefined}
+     */
+    _dfUnregisterChild(toRemove) {
+        const filtered = this.children.filter(child => child !== toRemove);
```
Why passing by a temp variable ?
```
this.children = this.children.filter(child => child !== toRemove);
```

DavidPineauScality replied:
Good point.
 - [ ] set `this.children` without intermediate

alexandre-merle commented on tests/unit/Logger.js:
```diff
@@ -256,3 +256,81 @@ describe('Werelogs Module-level Logger can log as specified by the log level', (
+            assert.strictEqual(requestInfo.object,
+                               dummyLogger.ops[0][1][0].object);
+            assert.strictEqual(requestInfo.creator,
+                               dummyLogger.ops[0][1][0].creator);
```
you can also check the name and the data here :) to check all default fields

DavidPineauScality replied:
- [ ] Check `name` and `data` fields here too

alexandre-merle commented on tests/unit/DefaultFields.js:
```diff
@@ -0,0 +1,500 @@
+        it('Can set a parent to a node', done => {
+            const childNode = linkedChild(parentNode);
+            assert.strictEqual(childNode.parent, parentNode);
+            assert.notEqual(parentNode.children.indexOf(childNode), -1);
```
minor: notStrictEqual :)

DavidPineauScality replied:
- [ ] Use `notStrictEqual` instead of `notEqual`

MathieuCassagne commented on lib/Logger.js:
```diff
@@ -77,13 +86,15 @@ class Logger {
     _doLog(levelName, msg, data) {
         const sLogger = Config.logger;
-        const finalData = { name: this.name, time: Date.now() };
+        const finalData = objectCopy({},
```
Instead of doing this objectCopy each time for the "time" field, cannot we add it to the defaults fields and only update it's value here ?

DavidPineauScality replied:
That would mean updating the default value of one logger element, thus updating all of its children logger default fields as a side-effect for EACH logging request. This would be so much more consuming than only doing an object copy at the time of the request. Imagine updating all of the precomputedFields of all RequestLoggers whenever you log something with the root module...
Also, that's how the RequestLogger has always been working, no reason to change ways when applying the scheme to the ModuleLogger

ThibaultRiviere replied:
As discuss offline the initialization of finalData will be move after the next check avoiding useless computation.
After that it's a :+1:

alexandre-merle reviewed (commented):
:)

DavidPineauScality commented:
@ThibaultRiviere @alexandre-merle @MathieuCassagne Some additional fixes + one very small feature were added in latest commits. please review :)

alexandre-merle reviewed (dismissed):
LGTM :) let's try this :)

alexandre-merle commented:
@ironman-machine try :D

ironman-machine commented:
Hello @alexandre-merle

Starting end to end procedure using the following payload:
{"REPO_NAME": "werelogs", "SCALITY_WERELOGS_BRANCH": "dev/ModuleHierarchy", "DEFAULT_BRANCH": "master"}

Please follow http://ci.ironmann.io/gh/scality/Integration/8068 for CI status.

DavidPineauScality commented:
E2E succeeded. Waiting for a last review, since @ThibaultRiviere is away at th moment, and thus cannot review the last bits yet.

DavidPineauScality commented:
@ironman-machine

ThibaultRiviere reviewed (dismissed):
Still ok!

ironman-machine commented:
PR has been updated. Reviewers, please be cautious.

lamphamsy commented on lib/DefaultFields.js:
```diff
@@ -0,0 +1,227 @@
+     * @return {undefined}
+     */
+    _dfUnregisterChild(toRemove) {
+        this.children = this.children.filter(child => child !== toRemove);
```
This action might take times. If it could be optimised, it can improve perf. Particularly, it is called from `setParent` that might occur frequently :)

DavidPineauScality replied:
Good point; our current attempt is extensively using the setParent function, incidentally.

Any suggestion is welcome. I mostly focused on having a proper design when implementing this, and expected a number of items to be resource-consuming and perf-consuming.

lamphamsy reviewed (commented):
LGTM, only a small comment :)

ghost commented:
project dropped
